### PR TITLE
refresh repository README on Docker hub after image push

### PR DESCRIPTION
**Pull request by vladak**
The README file on Docker hub is not automatically refreshed. This is because the Docker image builds are performed in Travis and not on Docker hub.

Using https://stackoverflow.com/a/55082686/11582827 and https://success.docker.com/article/how-do-i-authenticate-with-the-v2-api I made this work without enabling automated builds in Docker hub. The API does not seem to be publicly documented but works.

### CODE REVIEW

File: dev/docker.sh [hunk 1]
```diff
@@ -13,50 +13,15 @@
...
+IMAGE="opengrok/docker"
 
 # Travis can only work on master since it needs encrypted variables.
 if [ "${TRAVIS_PULL_REQUEST}" != "false" ]; then
-	echo "Not publishing docker image for pull requests"
+	echo "Not building docker image for pull requests"
 	exit 0
 fi
 
-# The push only works on the main repository.
```
Comment (tulinkry) [inline dev/docker.sh]: why is this no longer needed?

File: dev/docker.sh [hunk 2]
```diff
@@ -72,14 +37,77 @@ if [[ -z $DOCKER_PASSWORD ]]; then
...
+		echo "Successfully pushed README to Docker Hub"
+	else
+		printf "Unable to push README to Docker Hub, response code: %s\n" "${code}"
+		exit 1
+	fi
+}
+
+TOKEN=$(curl -s -H "Content-Type: application/json" -X POST \
+    -d '{"username": "'${DOCKER_USERNAME}'", "password": "'${DOCKER_PASSWORD}'"}' \
+    https://hub.docker.com/v2/users/login/ | jq -r .token)
```
Comment (tulinkry) [inline dev/docker.sh]: do you ensure jq is installed?

File: dev/docker.sh [hunk 3]
```diff
@@ -72,14 +37,77 @@ if [[ -z $DOCKER_PASSWORD ]]; then
...
+if [[ -z $TOKEN ]]; then
+	echo "Cannot get auth token"
```
Comment (tulinkry) [inline dev/docker.sh]: Cannot get auth token to publish the readme (so as you know in the log why it needs the token)

File: dev/docker.sh [hunk 4]
```diff
@@ -72,14 +37,77 @@ if [[ -z $DOCKER_PASSWORD ]]; then
 	exit 1
 fi
 
+VERSION="$TRAVIS_TAG"
+VERSION_SHORT=$( echo $VERSION | cut -d. -f1,2 )
```
Comment (tulinkry) [inline dev/docker.sh]: how about default "latest" ? when you build the image in your local dev env

File: dev/docker.sh [hunk 5]
```diff
@@ -72,14 +37,77 @@ if [[ -z $DOCKER_PASSWORD ]]; then
...
+		exit 1
+	fi
+
+	local code=$(jq -n --arg msg "$(<${input_file})" \
+	    '{"registry":"registry-1.docker.io","full_description": $msg }' | \
+	        curl -s -o /dev/null  -L -w "%{http_code}" \
+	           https://cloud.docker.com/v2/repositories/"${image}"/ \
+	           -d @- -X PATCH \
+	           -H "Content-Type: application/json" \
+	           -H "Authorization: JWT ${token}")
```
Comment (tulinkry) [inline dev/docker.sh]: just to understand it...this reads the input file (=README) and sends it through the api to docker, right?

Comment (tulinkry) [reply dev/docker.sh]: looks like nice oneliner but little hard to follow at first sight

Comment (vladak) [reply dev/docker.sh]: yes. I copied the whole thing from elsewhere.

File: dev/docker.sh (hunk 1 above)
Comment (vladak) [reply dev/docker.sh]: mismerge, should be fixed now.

File: dev/docker.sh (hunk 4 above)
Comment (vladak) [reply dev/docker.sh]: fixed the mismerge

File: dev/before_install [hunk 6]
```diff
@@ -20,7 +20,8 @@ if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
 	    python3 \
 	    python3-venv \
 	    python3-pip \
-	    nodejs
+	    nodejs \
+	    jq
```
Comment (tulinkry) [inline dev/before_install]: just curious...why not in the osx branch?

Comment (vladak) [reply dev/before_install]: just forget, adding.
